src/target/puya.c

```diff
@@ -0,0 +1,171 @@
+	flash->prepare = puya_flash_prepare;
+	flash->done = puya_flash_done;
+	flash->erased = 0xffU;
+	target_add_flash(target, flash);
+
+	return false;
+}
+
+static bool puya_flash_prepare(target_flash_s *flash)
+{
+	target_mem32_write32(flash->t, FLASH_KEYR, FLASH_KEYR_KEY1);
+	target_mem32_write32(flash->t, FLASH_KEYR, FLASH_KEYR_KEY2);
+
+	int hsi_fs = (target_mem32_read32(flash->t, RCC_ICSCR) >> 13) & 7;
+	if (hsi_fs > 4)
+		hsi_fs = 0;
+	DEBUG_INFO("HSI frequencey selection is %d\n", hsi_fs);
+
+	const uint32_t eppara0 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 0);
+	const uint32_t eppara1 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 4);
+	const uint32_t eppara2 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 8);
+	const uint32_t eppara3 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 12);
+	const uint32_t eppara4 = target_mem32_read32(flash->t, FLASH_TIMING_CAL_BASE + hsi_fs * 20 + 16);
+	DEBUG_INFO("PY32 flash timing cal 0: %08" PRIx32 "\n", eppara0);
+	DEBUG_INFO("PY32 flash timing cal 1: %08" PRIx32 "\n", eppara1);
+	DEBUG_INFO("PY32 flash timing cal 2: %08" PRIx32 "\n", eppara2);
+	DEBUG_INFO("PY32 flash timing cal 3: %08" PRIx32 "\n", eppara3);
+	DEBUG_INFO("PY32 flash timing cal 4: %08" PRIx32 "\n", eppara4);
+
+	target_mem32_write32(flash->t, FLASH_TS0, eppara0 & 0xff);
+	target_mem32_write32(flash->t, FLASH_TS1, (eppara0 >> 16) & 0x1ff);
+	target_mem32_write32(flash->t, FLASH_TS3, (eppara0 >> 8) & 0xff);
+	target_mem32_write32(flash->t, FLASH_TS2P, eppara1 & 0xff);
+	target_mem32_write32(flash->t, FLASH_TPS3, (eppara1 >> 16) & 0x7ff);
+	target_mem32_write32(flash->t, FLASH_PERTPE, eppara2 & 0x1ffff);
+	target_mem32_write32(flash->t, FLASH_SMERTPE, eppara3 & 0x1ffff);
+	target_mem32_write32(flash->t, FLASH_PRGTPE, eppara4 & 0xffff);
+	target_mem32_write32(flash->t, FLASH_PRETPE, (eppara4 >> 16) & 0x3fff);
+
+	return true;
+}
+
+static bool puya_flash_done(target_flash_s *flash)
+{
+	target_mem32_write32(flash->t, FLASH_CR, FLASH_CR_LOCK);
+	return true;
+}
+
+static bool puya_wait_flash(target_s *const target, platform_timeout_s *const timeout)
+{
+	while (target_mem32_read32(target, FLASH_SR) & FLASH_SR_BSY) {
+		if (target_check_error(target))
+			return false;
+		if (timeout)
+			target_print_progress(timeout);
+	}
+	return true;
+}
+
+static bool puya_flash_erase(target_flash_s *flash, target_addr_t addr, size_t len)
+{
+	(void)len;
+	DEBUG_INFO("PY32 flash erase @ %08" PRIx32 "\n", addr);
+	target_mem32_write32(flash->t, FLASH_CR, FLASH_CR_PER);
+	target_mem32_write32(flash->t, addr, 0);
+	return puya_wait_flash(flash->t, NULL);
+}
+
+static bool puya_flash_write(target_flash_s *flash, target_addr_t dest, const void *src, size_t len)
+{
+	(void)len;
+	DEBUG_INFO("PY32 flash write @ %08" PRIx32 "\n", dest);
+	target_mem32_write32(flash->t, FLASH_CR, FLASH_CR_PG);
+	for (uint32_t i = 0; i < FLASH_PAGE_SIZE; i += 4) {
```

Comment:
This is not guaranteed to get `FLASH_PAGE_SIZE` bytes passed to it - `len` is allowed to be `0 < len <= FLASH_PAGE_SIZE`. It is an error to access `src` past `len` bytes.

Reply:
I fixed this to support an arbitrary len (including unaligned non-multiple-of-4), although this is contrary to the behavior I actually observe and also goes against the assumption made in https://github.com/blackmagic-debug/blackmagic/blob/main/src/target/msp432e4.c#L287-L292 (which is where I copied this from).

Reply:
Indeed, the code in target_flash.c only ever calls write with writesize bytes: https://github.com/blackmagic-debug/blackmagic/blob/main/src/target/target_flash.c#L206

Reply:
Argh, ok, you are correct there - our memory was faulty. We'll change the request then to: rather than using the FLASH_PAGE_SIZE hard-coded constant here, please either use `flash->writesize`, or `len` to guard against both other people reading it and making the same mistake, and also in case the write size is changed at some later date so it can't get out of sync.

Don't worry about the other half of our original review note as the target Flash layer is preventing that being a problem.